controllers/route/route_certificaterequest_controller.go

```diff
@@ -0,0 +1,343 @@
+	cmutil "github.com/jetstack/cert-manager/pkg/api/util"
+	cmapi "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1"
+	cmmeta "github.com/jetstack/cert-manager/pkg/apis/meta/v1"
+	cmutilpki "github.com/jetstack/cert-manager/pkg/util/pki"
+	routev1 "github.com/openshift/api/route/v1"
+	outils "github.com/redhat-cop/operator-utils/pkg/util"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// CertificateRequestReconciler is a controller that creates cert-manager CertificateRequests
+// when it sees a route that contains cert-manager.io/issuer-name and cert-manager.io/issuer-kind
+// annotations.
+//
+// After the CertificateRequests have been signed, it updates the Route.Spec.TLSConfig block with
+// the generated private key and signed certificate. If the cert-manager Issuer supports sending relevant
+// CAs, it will also overwrite the TLSConfig.CA Certificate.
+//
+// The controller will renew the Certificate after 2/3 of the duration of the signed certificate passes.
+type CertificateRequestReconciler struct {
+	outils.ReconcilerBase
+	Log logr.Logger
+}
+
+func (c *CertificateRequestReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	isCertManagerRoute := predicate.NewPredicateFuncs(
+		func(object client.Object) bool {
+			route, ok := object.(*routev1.Route)
+			if !ok {
+				return false
+			}
+			return hasCertManagerAnnotations(route)
+		},
+	)
+
+	// Only process CRs that have been approved or denied
+	shouldProcessCR := predicate.NewPredicateFuncs(
+		func(object client.Object) bool {
+			cr, ok := object.(*cmapi.CertificateRequest)
+			if !ok {
+				return false
+			}
+			return cmutil.CertificateRequestIsApproved(cr) || cmutil.CertificateRequestIsDenied(cr)
+		},
+	)
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&routev1.Route{
+			TypeMeta: metav1.TypeMeta{Kind: "Route"},
+		}, builder.WithPredicates(isCertManagerRoute)).
+		// We will create cert-manager CertificateRequests to populate TLS routes, so
+		// we need to respond to the creation / update events for them too.
+		Watches(&source.Kind{
+			Type: &cmapi.CertificateRequest{
+				TypeMeta: metav1.TypeMeta{Kind: "CertificateRequest"},
+			},
+		}, &handler.EnqueueRequestForOwner{
+			OwnerType: &routev1.Route{},
+		}, builder.WithPredicates(shouldProcessCR)).
+		Complete(c)
+}
+
+func (c *CertificateRequestReconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	// Fetch the Route being reconciled
+	var instance *routev1.Route
+	err := c.GetClient().Get(ctx, req.NamespacedName, instance)
+	if err != nil {
+		return c.ManageError(ctx, instance, err)
```

Comment:
I am getting a segfault here (see below logs). Maybe `ManagedError` [does not seem](https://github.com/redhat-cop/operator-utils/blob/b22e79c476d3c075953718c8adbece4a0ae1c9f5/pkg/util/reconciler.go#L351
) to be able to handle nils, the function documentation does not say anything about nils. :disappointed: 

The segfault:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1375842]

goroutine 736 [running]:
github.com/openshift/api/route/v1.(*Route).GetObjectKind(0x0)
        <autogenerated>:1 +0x22
k8s.io/client-go/tools/reference.GetReference(0x120a0ea, {0x19107c8, 0x0})
        /home/mvalais/go/pkg/mod/k8s.io/client-go@v0.22.2/tools/reference/ref.go:59 +0x132
k8s.io/client-go/tools/record.(*recorderImpl).generateEvent(0xc000bd95c0, {0x19107c8, 0x0}, 0x120a745, {0x170ba69, 0x40b5cd}, {0x171385c, 0xf}, {0xc004e2ee80, 0x1d})
        /home/mvalais/go/pkg/mod/k8s.io/client-go@v0.22.2/tools/record/event.go:327 +0x76
k8s.io/client-go/tools/record.(*recorderImpl).Event(0x1127f59, {0x19107c8, 0x0}, {0x170ba69, 0x40c62a}, {0x171385c, 0x16dddc0}, {0xc004e2ee80, 0x0})
        /home/mvalais/go/pkg/mod/k8s.io/client-go@v0.22.2/tools/record/event.go:352 +0x4f
sigs.k8s.io/controller-runtime/pkg/internal/recorder.(*lazyRecorder).Event(0xc000a89b80, {0x19107c8, 0x0}, {0x170ba69, 0x7}, {0x171385c, 0xf}, {0xc004e2ee80, 0x1d})
        /home/mvalais/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.10.1/pkg/internal/recorder/recorder.go:155 +0xcb
github.com/redhat-cop/operator-utils/pkg/util.(*ReconcilerBase).ManageErrorWithRequeue(0xc000133860, {0x19248d8, 0xc004e6c840}, {0x196ccc0, 0x0}, {0x18f9920, 0xc004e73df0}, 0x196ccc0)
        /home/mvalais/go/pkg/mod/github.com/redhat-cop/operator-utils@v1.1.4/pkg/util/reconciler.go:380 +0xff
github.com/redhat-cop/operator-utils/pkg/util.(*ReconcilerBase).ManageError(...)
        /home/mvalais/go/pkg/mod/github.com/redhat-cop/operator-utils@v1.1.4/pkg/util/reconciler.go:407
github.com/redhat-cop/cert-utils-operator/controllers/route.(*CertificateRequestReconciler).Reconcile(0xc000133860, {0x19248d8, 0xc004e6c840}, {{{0xc00069f700, 0xf}, {0xc00069f6f0, 0x5}}})
        /home/mvalais/code/jetstack/cert-utils-operator/controllers/route/route_certificaterequest_controller.go:106 +0x84d
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0xc0002e6b40, {0x19248d8, 0xc004e6c7b0}, {{{0xc00069f700, 0x1607940}, {0xc00069f6f0, 0xc000158140}}})
        /home/mvalais/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.10.1/pkg/internal/controller/controller.go:114 +0x222
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc0002e6b40, {0x1924830, 0xc0005d0e00}, {0x1599180, 0xc000a26040})
        /home/mvalais/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.10.1/pkg/internal/controller/controller.go:311 +0x2f2
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0002e6b40, {0x1924830, 0xc0005d0e00})
        /home/mvalais/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.10.1/pkg/internal/controller/controller.go:266 +0x205
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
        /home/mvalais/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.10.1/pkg/internal/controller/controller.go:227 +0x85
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2
        /home/mvalais/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.10.1/pkg/internal/controller/controller.go:223 +0x354
```